[0.11.0] unsupported derivative iterator type

InfluxDB version information:

```
InfluxDB v0.11.0.n1455177634 (git: master 8d7a4a9accba0ace785218a1d8fcc36540f4bca4, built 2016-02-11T08:02:04.633643)
```

Query:

```
SELECT derivative(pointReq, 1s) FROM _internal.monitor."write" WHERE time > now() - 5m
```

Panic:

```
panic: unsupported derivative iterator type: *influxql.integerSortedMergeIterator

goroutine 3434 [running]:
github.com/influxdata/influxdb/influxql.newDerivativeIterator(0x7fba280bf198, 0xc209d2c900, 0x7fba280be7a0, 0xc20bb96ab0, 0x1056640, 0x0, 0x0, 0xc20dec4c50, 0x1, 0x1, ...)
        /tmp/tmp.aWGBEz2ODw/src/github.com/influxdata/influxdb/influxql/call_iterator.go:806 +0x3fa
github.com/influxdata/influxdb/influxql.buildExprIterator(0x7fba280be7a0, 0xc20bb96ab0, 0x7fba280bede0, 0xc20fb31220, 0x7fba280be7a0, 0xc20bb96ab0, 0x1056640, 0x0, 0x0, 0xc20dec4c50, ...)
        /tmp/tmp.aWGBEz2ODw/src/github.com/influxdata/influxdb/influxql/select.go:316 +0x1c65
github.com/influxdata/influxdb/influxql.func·033(0x0, 0x0)
        /tmp/tmp.aWGBEz2ODw/src/github.com/influxdata/influxdb/influxql/select.go:140 +0x5cb
github.com/influxdata/influxdb/influxql.buildFieldIterators(0xc20802b3c8, 0x1, 0x1, 0x7fba280bede0, 0xc20fb31220, 0x0, 0x0, 0x1056640, 0x0, 0x0, ...)
        /tmp/tmp.aWGBEz2ODw/src/github.com/influxdata/influxdb/influxql/select.go:175 +0x103
github.com/influxdata/influxdb/influxql.Select(0xc20f349520, 0x7fba280bede0, 0xc20fb31220, 0x0, 0x0, 0x0, 0x0, 0x0)
        /tmp/tmp.aWGBEz2ODw/src/github.com/influxdata/influxdb/influxql/select.go:67 +0xd6d
github.com/influxdata/influxdb/tsdb.(*QueryExecutor).PlanSelect(0xc20800ad70, 0xc20f349520, 0x2710, 0x0, 0x0, 0x0, 0x0)
        /tmp/tmp.aWGBEz2ODw/src/github.com/influxdata/influxdb/tsdb/query_executor.go:275 +0x6ff
github.com/influxdata/influxdb/tsdb.(*QueryExecutor).planStatement(0xc20800ad70, 0x7fba280be8d0, 0xc20f349520, 0xc20fb2fd0e, 0x9, 0x2710, 0x0, 0x0, 0x0, 0x0)
        /tmp/tmp.aWGBEz2ODw/src/github.com/influxdata/influxdb/tsdb/query_executor.go:524 +0x1a4
github.com/influxdata/influxdb/tsdb.(*QueryExecutor).executeStatement(0xc20800ad70, 0x0, 0x7fba280be8d0, 0xc20f349520, 0xc20fb2fd0e, 0x9, 0xc20a058d80, 0x2710, 0xc20a058d20, 0x0, ...)
        /tmp/tmp.aWGBEz2ODw/src/github.com/influxdata/influxdb/tsdb/query_executor.go:648 +0x8f
github.com/influxdata/influxdb/tsdb.func·004()
        /tmp/tmp.aWGBEz2ODw/src/github.com/influxdata/influxdb/tsdb/query_executor.go:172 +0xa1d
created by github.com/influxdata/influxdb/tsdb.(*QueryExecutor).ExecuteQuery
        /tmp/tmp.aWGBEz2ODw/src/github.com/influxdata/influxdb/tsdb/query_executor.go:227 +0x242
```

/cc @jsternberg @benbjohnson

--- a/influxql/call_iterator.go
+++ b/influxql/call_iterator.go
@@ -804,6 +804,8 @@ func newDerivativeIterator(input Iterator, opt IteratorOptions, interval Interva
 	switch input := input.(type) {
 	case FloatIterator:
 		return &floatReduceSliceIterator{input: newBufFloatIterator(input), opt: opt, fn: newFloatDerivativeReduceSliceFunc(interval, isNonNegative)}
+	case IntegerIterator:
+		return &integerReduceSliceFloatIterator{input: newBufIntegerIterator(input), opt: opt, fn: newIntegerDerivativeReduceSliceFunc(interval, isNonNegative)}
 	default:
 		panic(fmt.Sprintf("unsupported derivative iterator type: %T", input))
 	}
@@ -831,6 +833,48 @@ func newFloatDerivativeReduceSliceFunc(interval Interval, isNonNegative bool) fl
 			// Calculate the derivative of successive points by dividing the
 			// difference of each value by the elapsed time normalized to the interval.
 			diff := p.Value - prev.Value
+			elapsed := p.Time - prev.Time
+
+			value := 0.0
+			if elapsed > 0 {
+				value = diff / (float64(elapsed) / float64(interval.Duration))
+			}
+
+			prev = *p
+
+			// Drop negative values for non-negative derivatives.
+			if isNonNegative && diff < 0 {
+				continue
+			}
+
+			output = append(output, FloatPoint{Time: p.Time, Value: value})
+		}
+		return output
+	}
+}
+
+// newIntegerDerivativeReduceSliceFunc returns the derivative value within a window.
+func newIntegerDerivativeReduceSliceFunc(interval Interval, isNonNegative bool) integerReduceSliceFloatFunc {
+	prev := IntegerPoint{Time: -1}
+
+	return func(a []IntegerPoint, opt *reduceOptions) []FloatPoint {
+		if len(a) == 0 {
+			return []FloatPoint{}
+		} else if len(a) == 1 {
+			return []FloatPoint{{Time: a[0].Time, Nil: true}}
+		}
+
+		if prev.Time == -1 {
+			prev = a[0]
+		}
+
+		output := make([]FloatPoint, 0, len(a)-1)
+		for i := 1; i < len(a); i++ {
+			p := &a[i]
+
+			// Calculate the derivative of successive points by dividing the
+			// difference of each value by the elapsed time normalized to the interval.
+			diff := float64(p.Value - prev.Value)
 			elapsed := p.Time - prev.Time
 
 			value := 0.0

--- a/influxql/select_test.go
+++ b/influxql/select_test.go
@@ -1474,3 +1474,51 @@ func TestSelect_ParenExpr(t *testing.T) {
 		t.Fatalf("unexpected points: %s", spew.Sdump(a))
 	}
 }
+
+func TestSelect_Derivative_Float(t *testing.T) {
+	var ic IteratorCreator
+	ic.CreateIteratorFn = func(opt influxql.IteratorOptions) (influxql.Iterator, error) {
+		return &FloatIterator{Points: []influxql.FloatPoint{
+			{Name: "cpu", Time: 0 * Second, Value: 20},
+			{Name: "cpu", Time: 4 * Second, Value: 10},
+			{Name: "cpu", Time: 8 * Second, Value: 19},
+			{Name: "cpu", Time: 12 * Second, Value: 3},
+		}}, nil
+	}
+
+	// Execute selection.
+	itrs, err := influxql.Select(MustParseSelectStatement(`SELECT derivative(value, 1s) FROM cpu WHERE time >= '1970-01-01T00:00:00Z' AND time < '1970-01-01T00:00:16Z'`), &ic)
+	if err != nil {
+		t.Fatal(err)
+	} else if a := Iterators(itrs).ReadAll(); !deep.Equal(a, [][]influxql.Point{
+		{&influxql.FloatPoint{Name: "cpu", Time: 4 * Second, Value: -2.5}},
+		{&influxql.FloatPoint{Name: "cpu", Time: 8 * Second, Value: 2.25}},
+		{&influxql.FloatPoint{Name: "cpu", Time: 12 * Second, Value: -4}},
+	}) {
+		t.Fatalf("unexpected points: %s", spew.Sdump(a))
+	}
+}
+
+func TestSelect_Derivative_Integer(t *testing.T) {
+	var ic IteratorCreator
+	ic.CreateIteratorFn = func(opt influxql.IteratorOptions) (influxql.Iterator, error) {
+		return &IntegerIterator{Points: []influxql.IntegerPoint{
+			{Name: "cpu", Time: 0 * Second, Value: 20},
+			{Name: "cpu", Time: 4 * Second, Value: 10},
+			{Name: "cpu", Time: 8 * Second, Value: 19},
+			{Name: "cpu", Time: 12 * Second, Value: 3},
+		}}, nil
+	}
+
+	// Execute selection.
+	itrs, err := influxql.Select(MustParseSelectStatement(`SELECT derivative(value, 1s) FROM cpu WHERE time >= '1970-01-01T00:00:00Z' AND time < '1970-01-01T00:00:16Z'`), &ic)
+	if err != nil {
+		t.Fatal(err)
+	} else if a := Iterators(itrs).ReadAll(); !deep.Equal(a, [][]influxql.Point{
+		{&influxql.FloatPoint{Name: "cpu", Time: 4 * Second, Value: -2.5}},
+		{&influxql.FloatPoint{Name: "cpu", Time: 8 * Second, Value: 2.25}},
+		{&influxql.FloatPoint{Name: "cpu", Time: 12 * Second, Value: -4}},
+	}) {
+		t.Fatalf("unexpected points: %s", spew.Sdump(a))
+	}
+}

